akka-http-core/src/main/scala/akka/http/impl/engine/client/ProxyGraphStage.scala

```diff
@@ -0,0 +1,119 @@
+  extends GraphStage[BidiShape[ByteString, ByteString, ByteString, ByteString]] {
+
+  import ProxyGraphStage._
+
+  val bytesIn: Inlet[ByteString] = Inlet("OutgoingTCP.in")
+  val bytesOut: Outlet[ByteString] = Outlet("OutgoingTCP.out")
+
+  val sslIn: Inlet[ByteString] = Inlet("OutgoingSSL.in")
+  val sslOut: Outlet[ByteString] = Outlet("OutgoingSSL.out")
+
+  override def shape: BidiShape[ByteString, ByteString, ByteString, ByteString] = BidiShape.apply(sslIn, bytesOut, bytesIn, sslOut)
+
+  private val connectMsg = ByteString(s"CONNECT ${targetHostName}:${targetPort} HTTP/1.1\r\nHost: ${targetHostName}\r\n\r\n")
+
+  @scala.throws[Exception](classOf[Exception])
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new GraphStageLogic(shape) {
+    private var state: State = Starting
+
+    val parser = new HttpResponseParser(settings.parserSettings, HttpHeaderParser(settings.parserSettings, log)()) {
+      var first = true
+      override def handleInformationalResponses = false
+      override protected def parseMessage(input: ByteString, offset: Int): StateResult = {
+        if (first) {
+          first = false
+          super.parseMessage(input, offset)
+        } else {
+          emit(RemainingBytes(input.drop(offset)))
+          terminate()
+        }
+      }
+    }
+    parser.setContextForNextResponse(HttpResponseParser.ResponseContext(HttpMethods.CONNECT, None))
+
+    setHandler(sslIn, new InHandler {
+      override def onPush() = {
+        state match {
+          case Starting   ⇒
+          // should not happen
+          case Connecting ⇒
+          // should not happen
+          case Connected ⇒
+            push(bytesOut, grab(sslIn))
+        }
+      }
+    })
+
+    setHandler(bytesIn, new InHandler {
+      override def onPush() = {
+        state match {
+          case Starting ⇒
+          // should not happen to be in Starting state
+          case Connecting ⇒
+            val proxyResponse = grab(bytesIn)
+            parser.parseBytes(proxyResponse) match {
+              case NeedMoreData ⇒ pull(bytesIn)
+              case ResponseStart(StatusCodes.OK, _, _, _, _) ⇒
+                state = Connected
+                pull(sslIn)
+                pull(bytesIn)
+              case ResponseStart(statusCode, _, _, _, _) ⇒
+                throw new IllegalStateException(s"unexpected proxy response status code: ${statusCode.intValue}")
+              case other ⇒
+                throw new IllegalStateException(s"unexpected element of type ${other.getClass}")
+            }
+
+          case Connected ⇒
+            push(sslOut, grab(bytesIn))
+        }
+      }
+    })
+
+    setHandler(bytesOut, new OutHandler {
+      override def onPull() = {
+        state match {
+          case Starting ⇒
+            push(bytesOut, connectMsg)
+            state = Connecting
+          case Connecting ⇒
```

Comment:
Fixed that, and added test in https://github.com/akka/akka-http/pull/921/files#diff-f2ab7fa3256d9d613bdc4d73f5479281R21 - fails with previous version of code